packages/excalidraw/components/App.tsx

```diff
@@ -7993,12 +7993,35 @@ class App extends React.Component<AppProps, AppState> {
 
         if (newElement.type === "freedraw") {
           const points = newElement.points;
-          const dx = pointerCoords.x - newElement.x;
-          const dy = pointerCoords.y - newElement.y;
+          let dx = pointerCoords.x - newElement.x;
+          let dy = pointerCoords.y - newElement.y;
+
+          if(event.shiftKey) {
+
+            const angle = Math.atan2(dy, dx);
+            const distance = Math.sqrt(dx * dx + dy * dy);
+
+            const snappingAngles = [
+              0, Math.PI / 6, Math.PI / 4, Math.PI / 3, Math.PI / 2,
+              (2 * Math.PI) / 3, (3 * Math.PI) / 4, (5 * Math.PI) / 6,
+              Math.PI, (7 * Math.PI) / 6, (5 * Math.PI) / 4, (4 * Math.PI) / 3,
+              (3 * Math.PI) / 2, (5 * Math.PI) / 3, (7 * Math.PI) / 4, (11 * Math.PI) / 6
+            ];
+            
+            const closestAngle = snappingAngles.reduce((prev, curr) =>
+              Math.abs(curr - angle) < Math.abs(prev - angle) ? curr : prev
+            );
 
+            dx = Math.cos(closestAngle) * distance;
+            dy = Math.sin(closestAngle) * distance;
+          }
           const lastPoint = points.length > 0 && points[points.length - 1];
-          const discardPoint =
-            lastPoint && lastPoint[0] === dx && lastPoint[1] === dy;
+
+          const minDistance = 15;
+          const maxDistance = 200;
+          const distanceFromLastPoint = lastPoint ? Math.sqrt(Math.pow(lastPoint[0] - dx, 2) + Math.pow(lastPoint[1] - dy, 2)) : Infinity;
```

Comment:
There is a function for classic point distance measurement, `pointDistance` but in this case there is no reason to calculate the distance instead of the distance squared, which is `pointDistanceSq` (which is now incorrectly implemented, that could be fixed as well).